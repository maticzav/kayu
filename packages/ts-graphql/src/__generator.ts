/**
 * These exports are meant to be used solely by the generated
 * code. Do not use them anywhere. These functions are undocumented
 * and might change without breaking change notice.
 *
 * Instead, use functions exported from the index file.
 */
export { Argument, arg } from './document/argument'
export { Field, composite, fragment, leaf } from './document/field'
export { nullable, list } from './internal/decode'
export { Fields, SelectionSet, selection } from './selection'
