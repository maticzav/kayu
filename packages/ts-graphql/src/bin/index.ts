#!/usr/bin/env node

import * as fs from 'fs'
import { IntrospectionSchema } from 'graphql'
import * as path from 'path'
import {
  GQLGenerator,
  loadSchemaFromPath,
  loadSchemaFromURL,
} from 'ts-graphql-codegen'
import { promisify } from 'util'

/* Utils */

const writefile = promisify(fs.writeFile)
const readdir = promisify(fs.readdir)

/* Configuration */

const DEFAULT_CONFIG = (root: string) => `
module.exports = {
  endpoint: "http://localhost:3000",
  schema: "${root}/schema.json",
  api: "${root}/api.ts",
  // prettier: undefined
}
`

/* Types */

export type Config = {
  /**
   * GraphQL server endpoint.
   */
  endpoint: string
  /**
   * Path to output file relative to this file.
   */
  api: string
  /**
   * Path to codecs declaration relative to this file.
   */
  codecs?: string
  /**
   * Path to schema cache relative to this file.
   */
  schema: string
  /**
   * Optional path to prettier.config.js path relative to this file.
   */
  prettier?: string
}

/**
 * Handles generation of configuration.
 */
export async function generate(root: string) {
  /**
   * Calculate the config path based on the root.
   */
  const configpath = path.resolve(root, './ts-graphql.config.js')

  let config: Config

  try {
    config = require(configpath)
  } catch (err) {
    console.log(`Coudln't find configuration.`)
    await writefile(configpath, DEFAULT_CONFIG(root))
    console.log(`We have scaffolded config file to ${configpath}.`)
    process.exit(0)
  }

  /**
   * Create boilerplate configuration file.
   */

  /**
   * We calculate output paths.
   */

  const ENDPOINT = config.endpoint
  const CORE_PATH = 'ts-graphql/__generator'
  const API_PATH = path.resolve(root, config.api)
  const SCHEMA_PATH = path.resolve(root, config.schema)

  // Optional args
  let CODECS_PATH: string | undefined = undefined
  let PRETTIER: object = {}

  if (config.prettier) {
    const PRETTIER_PATH = path.resolve(root, config.prettier)
    try {
      PRETTIER = require(PRETTIER_PATH)
    } catch (err) {
      console.log(`Couldn't find prettier.config.js in ${PRETTIER_PATH}.`)
      process.exit(1)
    }
  }
  if (config.codecs) {
    CODECS_PATH = path.resolve(root, config.codecs)
  }

  /**
   * We load the schema from the server if server is available,
   * otherwise we use the cached result.
   */
  let schema: IntrospectionSchema | undefined = undefined

  try {
    schema = await loadSchemaFromURL(ENDPOINT)
    await writefile(SCHEMA_PATH, JSON.stringify(schema, null, 2))
  } catch (err) {
    schema = await loadSchemaFromPath(SCHEMA_PATH)
  } finally {
    if (schema === undefined) throw new Error(`Couldn't load schema.`)
  }

  /**
   * Construct the generator.
   */
  const generator = new GQLGenerator(schema, PRETTIER)

  /**
   * Save the code and perform tests as described above.
   */
  const code = generator.generate({
    core: CORE_PATH.replace(/\.ts$/, ''),
    codecs: CODECS_PATH?.replace(/\.ts$/, ''),
  })
  await writefile(API_PATH, code)

  console.log(`API generated!`)
}

/* Main */

async function main() {
  /**
   * We calculate endpoint relative to configuration file.
   */
  const root = await pkg(process.cwd())

  /* istanbul ignore if */
  if (root === null) {
    console.log(`Couldn't find root folder of your project.`)
    process.exit(1)
  }

  generate(root)
}

/**
 * Run the script when we run this file.
 */
if (require.main === module) {
  main()
}

/* Utility functions */

/**
 * Tries to find the first package.json up the way. We use it
 * as a relative cornerstone for calculating paths.
 */
export async function pkg(cwd: string): Promise<string | null> {
  let currentPath: string = cwd

  /**
   * We iterate over the path until we either find package.json
   * or reach root.
   */
  while (currentPath !== '/') {
    const files = await readdir(currentPath)

    if (files.some((file) => file === 'package.json')) {
      return currentPath
    }
  }

  /* istanbul ignore next */
  return null
}
