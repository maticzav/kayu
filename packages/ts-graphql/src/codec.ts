/**
 * This file contains a spec for codec "protocol"
 * that outlines how scalars are introduced to the library.
 *
 * We assume that types starting with a lowercase letter represent
 * TypeScript's built-in types and need no importing. Every other type
 * will be imported from a given file.
 */

type JSON = string | number | null | undefined | { [key: string]: JSON }
export interface Codec<T> {
  /**
   * TypeScript type that represents it.
   */
  type: string
  /**
   * The value that we use for mocking - it can be anything that represents your type.
   */
  mock: T
  /**
   * The function that we use to decode from JSON response to a value.
   */
  decode: (json: unknown) => T
  /**
   * Converts a scalar into a JSON representable value.
   */
  serialize: (val: T) => JSON
}
