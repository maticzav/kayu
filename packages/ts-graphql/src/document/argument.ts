import * as hasher from 'object-hash'

type InternalArgument = {
  name: string
  type: string
  value?: any
}

/**
 * Represents an argument in selection.
 * Value is internally represented as any serializable type.
 */
export class Argument {
  /* State */
  private arg: InternalArgument

  /* Initializer */

  constructor(arg: InternalArgument) {
    this.arg = arg
  }

  /* Accessors */

  /**
   * Returns the name of an argument.
   */
  get name(): string {
    return this.arg.name
  }

  /**
   * Returns the type of an argument.
   */
  get type(): string {
    return this.arg.type
  }

  /**
   * Returns the value of an argument.
   */
  get value(): any | undefined {
    return this.arg.value
  }

  /**
   * Returns the alias of an argument.
   * We use alias to reference arguemnts globally in the query.
   */
  get alias(): string {
    return `_${this.hash}`
  }

  /**
   * Calculates the hash of the argument.
   */
  get hash(): string {
    return hasher.MD5(this.arg)
  }

  /* Methods */

  /**
   * Returns the JSON represetnation of an argument.
   */
  toJSON(): object {
    return this.arg
  }
}

/**
 * Creates a new argument instance.
 */
export function arg<T>(name: string, type: string, value?: T): Argument {
  return new Argument({ name, type, value })
}
