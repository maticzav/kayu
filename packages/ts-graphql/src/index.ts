export { HttpHeaders, send } from './http'
export { Codec } from './codec'
