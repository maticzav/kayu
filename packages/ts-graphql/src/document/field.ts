import { Argument, hash } from './argument'

export type Field =
  | {
      kind: 'composite'
      name: string
      arguments: Argument[]
      selection: Field[]
    }
  | { kind: 'leaf'; name: string; arguments: Argument[] }
  | { kind: 'fragment'; type: string; selection: Field[] }

/**
 * Creates a new leaf field.
 */
export function leaf(name: string, args: Argument[] = []): Field {
  return { kind: 'leaf', name, arguments: args }
}

/**
 * Creates a new composite field.
 */
export function composite(
  name: string,
  selection: Field[],
  args: Argument[] = [],
): Field {
  return { kind: 'composite', name, arguments: args, selection }
}

/**
 * Creates a new fragment field.
 */
export function fragment(type: string, selection: Field[]): Field {
  return { kind: 'fragment', type, selection }
}

/* Utility functions */

/**
 * Returns an alias of a given field if it has arguments.
 * Otherwise, it returns undefined.
 */
export function alias(field: Field): string | undefined {
  switch (field.kind) {
    case 'leaf':
    case 'composite': {
      return `${field.name}_${hash(field.arguments)}`
    }
    case 'fragment': {
      return undefined
    }
  }
}

/**
 * Returns all arguments in fields' tree.
 */
export function argumentsOfFields(fields: Field[]): Argument[] {
  return fields.flatMap((field) => {
    /**
     * Reuse it for every field.
     */
    switch (field.kind) {
      case 'leaf': {
        return field.arguments
      }
      case 'composite': {
        return [...field.arguments, ...argumentsOfFields(field.selection)]
      }
      /**
       * Recursively gets arguments of a subselection.
       */
      case 'fragment': {
        return [...argumentsOfFields(field.selection)]
      }
    }
  })
}
