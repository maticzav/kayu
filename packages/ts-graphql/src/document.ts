import { Field } from './document/field'
import { indent, defined } from './utils'

export enum OperationType {
  Query = 'query',
  Mutation = 'mutation',
}

export type SerializeInput = {
  operationType: OperationType
  fields: Field[]
  operationName?: string
}

/**
 * Serializes an operation into a GraphQL SDL.
 */
export function serialize({
  operationType,
  fields,
  operationName,
}: SerializeInput): string {
  /**
   * Get all the arguments in a fields' tree and join them in a single
   * declaration for operation definition. We first get all the arguments,
   * then filter out the ones without value, and parse the remaining ones.
   */
  let args: string | undefined = undefined

  const serializedArgs = fields
    .flatMap((f) => f.arguments)
    .filter((arg) => defined(arg.value))
    .map((arg) => `$${arg.alias}: ${arg.type}`)

  if (serializedArgs.length > 0) {
    args = `(${serializedArgs.join(', ')})`
  }

  /**
   * Construct operation definition by filtering out the parts
   * that may be empty and joining them.
   */
  const operationDefinition = [operationType, operationName, args]
    .filter<string>(defined)
    .join(' ')

  /**
   * Construct the query using given data.
   */
  const query: string[] = [
    `${operationDefinition} {`,
    ...fields.flatMap((f) => f.serialize()).map(indent(2)),
    `}`,
  ]
  return query.join('\n')
}
