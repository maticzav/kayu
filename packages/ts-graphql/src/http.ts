import { fetch } from 'cross-fetch'

import { OperationType, serialize } from './document'
import { hash } from './document/argument'
import { argumentsOfFields } from './document/field'
import { Fields, SelectionSet } from './selection'
import { defined } from './utils'

/**
 * Executes a query against the endpoint.
 */
// export async function send<TypeLock, Type>(opts: {
//   selection: SelectionSet<TypeLock, Type>
//   operation: OperationType
//   operationName?: string
//   endpoint: string
//   method: HttpMethod
//   headers: HttpHeaders
// }): Promise<Type> {
//   return perform({
//     endpoint,
//   })
// }

export type HttpHeaders = { [key: string]: string }
export enum HttpMethod {
  POST = 'POST',
  GET = 'GET',
}

interface PerformInput<TypeLock, Type> {
  endpoint: string
  /**
   * Selection of fields.
   */
  selection: SelectionSet<TypeLock, Type>
  /**
   * Tells whether it's a query or a mutation.
   */
  operation: OperationType
  operationName?: string
  /**
   * HTTP method used to send the request. Defaults to POST>
   */
  method?: HttpMethod
  /**
   * Dictionary of key-value pairs that we should use as headers.
   */
  headers: HttpHeaders
}

/**
 * Performs a request against the server.
 *
 * NOTE: This function is only used internally as the ground for query and mutation send
 *       functions defined above.
 */
async function perform<TypeLock, Type>(
  opts: PerformInput<TypeLock, Type>,
): Promise<Type> {
  /* Construct a request. */
  let fields = new Fields<TypeLock>()
  opts.selection.decoder(fields)

  const query = serialize({
    fields: fields.selection,
    operationType: opts.operation,
    operationName: opts.operationName,
  })

  let variables: { [hash: string]: any } = {}
  for (const arg of argumentsOfFields(fields.selection)) {
    if (defined(arg.value)) {
      variables[hash([arg])] = arg.value
    }
  }

  const body: { [key: string]: any } = {
    query,
    variables,
  }

  if (opts.operationName) {
    body['operationName'] = opts.operationName
  }

  /* Perform a request. */
  const res = await fetch(opts.endpoint, {
    method: opts.method || 'POST',
    body: JSON.stringify(body),
    headers: {
      ...opts.headers,
      'Content-Type': 'application/json',
    },
  })

  /* Decode the result. */
  const data = await res.json()
  const fieldsWithData = new Fields<TypeLock>(data)

  return opts.selection.decoder(fieldsWithData)
}
