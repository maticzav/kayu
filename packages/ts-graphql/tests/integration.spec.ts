import * as express from 'express'
import { Server } from 'http'

import { apollo } from '../../../server/src'

import { serialize, OperationType } from '../src/document'
import { send } from '../src'

import * as api from './__fixtures__/api'

/* Config */

const PORT = 4321

/* Tests */

describe('integration', () => {
  /* State */

  let url = `http://localhost:${PORT}/graphql`
  let server: Server

  /**
   * We first start the server.
   */
  beforeAll(async () => {
    const app = express()
    apollo.applyMiddleware({ app })
    server = app.listen(PORT)
  })

  /**
   * We finish by shutting down the server.
   */
  afterAll(async () => {
    server.close()
  })

  /**
   * Tests fleet.
   */
  test('tests the API', async () => {
    const pg_human = api.objects.human((t) => {
      let id = t.id()
      let name = t.name()
      let appearsIn = t.appearsIn()

      return { id, name, appearsIn }
    })

    const pg = api.objects.query((t) => {
      let human = t.human({ id: '1000' })(pg_human.nullable)
      return { human }
    })

    const res = await send({
      endpoint: url,
      selection: pg,
    })

    expect(res).toMatchSnapshot()
  })
})
