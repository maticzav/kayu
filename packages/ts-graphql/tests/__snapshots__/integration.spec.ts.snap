// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration tests the API 1`] = `
Object {
  "human": Object {
    "appearsIn": Array [
      "NEWHOPE",
      "EMPIRE",
      "JEDI",
    ],
    "id": "1000",
    "name": "Luke Skywalker",
  },
}
`;
